_posts/2019-12-20-Our-Devoxx-Demo-App.md

```diff
@@ -0,0 +1,225 @@
+After we have a stream we fetch the `video` tag from the page, assign the stream to it and let it play.
+We also get our canvas in which we will display the actual video output, the video tag itself is hidden.
+The canvas allows us to annotate the image with a bounding box and some extra information about the detected objects.
+Finally we call the update method which will be called for each update.
+
+``` javascript
+async function update(video, canvas, context, detector) {
+    context.drawImage(video, 0, 0, canvas.width, canvas.height);
+
+    const detectedClasses = await detector.detect(canvas);
+    Loader.anotateCanvas(canvas, detectedClasses);
+    updateList(detectedClasses);
+
+    if (enableLiveUpdate) {
+        setTimeout(update, 1000 / framerate, video, canvas, context, detector);
+    }
+}
+```
+This function will function as our update loop/tick.
+We draw the image to the canvas, just as it was seen by the webcam.
+We then pass the canvas into the detector which will detect any objects in the image data contained in the canvas.
+This returns an array of detected classes which contain the type of objects, the score and a bounding box.
+That information is fed to a utility method that will annotate the canvas with a red rectangle with the coordinates of the bounding and the type of object with its score.
+The update function will be called again with a very basic (and very limited) framerate timeout.
+
+There is other code, mainly in the coco folder which is mostly taken from the [default google implementation](https://github.com/tensorflow/tfjs-models/tree/master/coco-ssd){:target="_blank" rel="noopener noreferrer"} and modified to match the needs of this application.
+One thing worth noting is the changes required to a small piece of code to make the model work faster by allowing it to work asynchronously:
+```typescript
+    const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);
+
+    const indexTensor = tf.tidy(() => {
+        const boxes2 = tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
+        return tf.image.nonMaxSuppression(boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
+    });
+
+    const indexes = indexTensor.dataSync() as Float32Array;
+    indexTensor.dispose();
+```
+
+Needs to be changed into:
+
+```typescript
+    const [maxScores, classes] = this.calculateMaxScores(scores, result[0].shape[1], result[0].shape[2]);
+
+    const boxes2 = tf.tensor2d(boxes, [result[1].shape[1], result[1].shape[3]]);
+    const temp = await tf.image.nonMaxSuppressionAsync(boxes2, maxScores, maxNumBoxes, 0.5, 0.5);
+    const indexTensor = tf.tidy(() => {
+        return temp;
+    });
+
+    const indexes = indexTensor.dataSync() as Float32Array;
+    indexTensor.dispose();
+```
+
+Because the `tf.image.nonMaxSuppressionAsync`returns a Promise the replacement is not as simple as just replacing the method call to the async variant.
```

Comment:
```suggestion
Because the `tf.image.nonMaxSuppressionAsync`returns a `Promise`, the replacement is not as simple as just replacing the method call to the async variant.
```